crates/bevy_ecs/src/world/testing_tools.rs

```diff
@@ -0,0 +1,96 @@
+use crate::component::Component;
+use crate::entity::Entity;
+use crate::event::Events;
+use crate::schedule::{Stage, SystemStage};
+use crate::system::{In, IntoChainSystem, IntoSystem};
+use crate::world::{FilterFetch, Mut, Resource, World, WorldQuery};
+use std::fmt::Debug;
+
+impl World {
+    /// Asserts that that the current value of the resource `R` is `value`
+    pub fn assert_resource_eq<R: Resource + PartialEq + Debug>(&self, value: R) {
+        let resource = self
+            .get_resource::<R>()
+            .expect("No resource matching the type of {value} was found in the world.");
+        assert_eq!(*resource, value);
+    }
+
+    /// Asserts that that the current value of the non-send resource `NS` is `value`
+    pub fn assert_nonsend_resource_eq<NS: 'static + PartialEq + Debug>(&self, value: NS) {
+        let resource = self
+            .get_non_send_resource::<NS>()
+            .expect("No non-send resource matching the type of {value} was found in the world.");
+        assert_eq!(*resource, value);
+    }
+
+    /// Asserts that all components of type `C` returned by a query with the filter `F` will equal `value`
+    ///
+    /// This is commonly used with the corresponding `query_len` method to ensure that the returned query is not empty.
+    ///
+    /// WARNING: because we are constructing the query from scratch,
+    /// [`Changed`](crate::query::Changed) and [`Added`](crate::query::Added) filters
+    /// will always return true.
+    pub fn assert_component_eq<C, F>(&mut self, value: &C)
+    where
+        C: Component + PartialEq + Debug,
+        F: WorldQuery,
+        <F as WorldQuery>::Fetch: FilterFetch,
+    {
+        let mut query_state = self.query_filtered::<(Entity, &C), F>();
+        for (entity, component) in query_state.iter(self) {
+            if component != value {
+                panic!(
+                    "Found component {component:?} for {entity:?}, but was expecting {value:?}."
+                );
+            }
+        }
+    }
+
+    /// Returns the number of entities found by the [`Query`](crate::system::Query) with the type parameters `Q` and `F`
+    pub fn query_len<Q, F>(&mut self) -> usize
+    where
+        Q: WorldQuery,
+        F: WorldQuery,
+        <F as WorldQuery>::Fetch: FilterFetch,
+    {
+        let mut query_state = self.query_filtered::<Q, F>();
+        query_state.iter(self).count()
+    }
+
+    /// Sends an `event` of type `E`
+    pub fn send_event<E: Resource>(&mut self, event: E) {
+        let mut events: Mut<Events<E>> = self.get_resource_mut()
+        .expect("The specified event resource was not found in the world. Did you forget to call `app.add_event::<E>()`?");
+
+        events.send(event);
+    }
+
+    /// Returns the number of events of the type `E` that were sent this frame
+    pub fn events_len<E: Resource + Debug>(&self) -> usize {
+        let events = self.get_resource::<Events<E>>().unwrap();
```

Comment:
I would change this to use `expect` instead of `unwrap` and add a message that the user maybe forgot to add the event using `app.add_event` (see `send_event`)